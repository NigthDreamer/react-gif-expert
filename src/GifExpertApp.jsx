import { useState } from "react";
import { AddCategory, GifGrid } from "./components";

const GifExpertApp = () => {
  const [categories, setCategories] = useState(["Dragon Ball"]);

  const onAddCategory = (newCategory) => {
    //No se puede mutar (editar directamente) categories del Hook
    // const updatedCategories = [...categories, newCategory];

    // setCategories(updatedCategories);

    // setCategories(cat => [...cat, newCategory]);

    if (!categories.includes(newCategory)) {
      setCategories([newCategory]);
    }
  };

  return (
    <>
      <h1>GifExpertApp</h1>

      <AddCategory
        // setCategories={ setCategories }
        onNewCategory={onAddCategory}
      />

      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  );
};

export default GifExpertApp;
