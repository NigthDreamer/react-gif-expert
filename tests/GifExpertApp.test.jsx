import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import GifExpertApp from "../src/GifExpertApp";

describe("Pruebas en <GifExpertApp/>", () => {
  test("Debe de hacer match con el snapshot", () => {
    const { container } = render(<GifExpertApp />);

    expect(container).toMatchSnapshot();
  });

  test("Al iniciar la aplicacion, se cargan las imagenes por defecto y se muestran", async () => {
    render(<GifExpertApp />);

    await waitFor(() =>
      expect(screen.getAllByRole("img").length).toBeGreaterThan(0)
    );

    const images = screen.getAllByRole("img");

    expect(images.length).toBeGreaterThan(0);
  });

  test("Al hacer submit de un nuevo valor en el input, se cargan las nuevas imagenes", async () => {
    render(<GifExpertApp />);

    await waitFor(() =>
      expect(screen.getAllByRole("img").length).toBeGreaterThan(0)
    );

    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    const newCategory = "Juego de Tronos";
    const oldImages = screen.getAllByRole("img");

    //Meto el valor en el input para setear el valor del formulario
    fireEvent.input(input, { target: { value: newCategory } });
    //Submiteo el valor del formulario para triggerear el hook y cargar las imagenes
    fireEvent.submit(form);

    await waitFor(() =>
      expect(screen.getAllByRole("img").length).toBeGreaterThan(0)
    );

    const newImages = screen.getAllByRole("img");

    expect(oldImages).not.toEqual(newImages);
  });
});
