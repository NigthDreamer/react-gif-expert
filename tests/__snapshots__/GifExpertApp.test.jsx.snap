// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifExpertApp/> Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    GifExpertApp
  </h1>
  <form
    aria-label="form"
  >
    <input
      id="cat-inp"
      placeholder="Buscar gifs"
      type="text"
      value=""
    />
  </form>
  <h3>
    Dragon Ball
  </h3>
  <h2>
    Cargando...
  </h2>
  <div
    class="card-grid"
  />
</div>
`;
